Improved performance of _index_for_assignment and issue fix

sytham opened the pull request:
- fixed problem with reduce() if variables are passed in in different
  order from which they are in factor scope. Fixes pgmpy/pgmpy#381
  (partial fix; chose not to fix problem 2)
- improved performance of _index_for_assignment: 4 - 50x faster. See below timeit results:

setup: from pgmpy.factors import Factor, factor_product
F = Factor(['1','2','3','4','5','6','7','8'], [4,2,2,3,5,2,2,3], range(2880))

timeit.timeit("F._index_for_assignment([1,1,1,1,1,1,1,1])", setup=setup, number=10000)
0.18317501399906178
timeit.timeit("F._index_for_assignment_old([1,1,1,1,1,1,1,1])", setup=setup, number=10000)
0.44581467299940414

timeit.timeit("F._index_for_assignment([1,1,-1,1,-1,1,1,1])", setup=setup, number=10000)
0.9840771199997107
timeit.timeit("F._index_for_assignment_old([1,1,-1,1,-1,1,1,1])", setup=setup, number=10000)
2.7516035469998315

timeit.timeit("F._index_for_assignment([1,1,-1,1,-1,1,-1,1])", setup=setup, number=10000)
1.382639981999091
timeit.timeit("F._index_for_assignment_old([1,1,-1,1,-1,1,-1,1])", setup=setup, number=10000)
4.924537786999281

timeit.timeit("F._index_for_assignment([1,1,-1,1,-1,1,-1,-1])", setup=setup, number=10000)
1.7535868900013156
timeit.timeit("F._index_for_assignment_old([1,1,-1,1,-1,1,-1,-1])", setup=setup, number=10000)
13.60733557200001

timeit.timeit("F._index_for_assignment([1,-1,-1,-1,-1,-1,-1,-1])", setup=setup, number=10000)
3.316042758000549
timeit.timeit("F._index_for_assignment_old([1,-1,-1,-1,-1,-1,-1,-1])", setup=setup, number=10000)
161.2918238050006

sytham committed:
- improved performance of _index_for_assignment: 4 - 50x faster
- fixed problem with reduce() if variables are passed in in different
order from which they are in factor scope. Fixes pgmpy/pgmpy#381
(partial fix; chose not to fix problem 2)

abinashpanda commented on pgmpy/factors/Factor.py:
```diff
@@ -4,6 +4,7 @@
 from collections import OrderedDict
 
 import numpy as np
+from sklearn.utils.extmath import cartesian
```
Please don't add any other dependencies unless we can't deal with it. `cartesian` method is not a very complicated method and can be implemented very easily in `pgmpy` as well. So adding `sklearn` as a dependency just for this method doesn't make any sense.

abinashpanda commented:
@sytham As per your comments in the PR, you said that you had made some changes in the `_index_for_assignment` method. But I don't see them in this PR. Could you please check that.

sytham commented:
If you check the diff on Factor.py, lines 410 - 422 in red are what I removed from `_index_for_assignment` and lines 414 - 430 are what I replaced it with. I essentially vectorized the code.

Sure, I can add the cartesian function in pgmpy itself. Where do you think it's best to add it? Create a utils package?

ankurankan commented:
Not relevant anymore since `_index_for_assignment` has been removed.
